### PR TITLE
[RFR] Add a serializer support

### PR DESCRIPTION
**Pull request by florianbelhomme**
* [x] Add tests
* [x] Add documentation

### CODE REVIEW

**Comment by florianbelhomme** [comment]
Niveau tech, il reste les classes & le contexte a bien gérer.
Pour les autres, il reste les tests et la documentation.

File: src/Serializer/SerializerAdapterInterface.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * Serializes data in the appropriate format.
+     *
+     * @param mixed       $data    any data
+     * @param string      $format  format name
+     * @param array|null  $context options normalizers/encoders have access to
+     *
+     * @return string
+     */
+    public function serialize($data, $format, array $context = null);
```
Comment (csarrazi) [inline src/Serializer/SerializerAdapterInterface.php]: Should rather be

```php
public function serialize($data, $format, array $context = array());
```

File: src/Serializer/SerializerAdapterInterface.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+     * Deserializes data into the given type.
+     *
+     * @param mixed       $data
+     * @param string      $type
+     * @param string      $format
+     * @param array|null  $context
+     *
+     * @return mixed
+     */
+    public function deserialize($data, $type, $format, array $context = null);
```
Comment (csarrazi) [inline src/Serializer/SerializerAdapterInterface.php]: Should rather be

```php
public function deserialize($data, $type, $format, array $context = array());
```

File: src/Serializer/JMSSerializerAdapter.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        $this->serializer = $serializer;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function serialize($data, $format, array $context = null)
+    {
+        if ($context !== null) {
+            $context = new JMSSerializationContext($context);
```
Comment (csarrazi) [inline src/Serializer/JMSSerializerAdapter.php]: You should rather use `SerializationContext::create()`.

File: src/Serializer/JMSSerializerAdapter.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        return $this->serializer->serialize($data, $format/*, $context*/);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function deserialize($data, $type, $format, array $context = null)
+    {
+        if ($context !== null) {
+            $context = new JMSDeserializationContext($context);
```
Comment (csarrazi) [inline src/Serializer/JMSSerializerAdapter.php]: You should rather use `SerializationContext::create()`.

File: src/GuzzleHttp/Cache/DoctrineAdapter.php [hunk 5]
```diff
@@ -43,7 +43,7 @@ public function save(RequestInterface $request, ResponseInterface $response)
         $data = [
             'status' => $response->getStatusCode(),
             'headers' => $response->getHeaders(),
-            'body' => (string)$response->getBody(),
+            'body' => (string) $response->getBody(),
```
Comment (csarrazi) [inline src/GuzzleHttp/Cache/DoctrineAdapter.php]: This change should be reverted. We don't want to pollute this PR with CS-related fixes.

File: src/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -74,6 +74,7 @@ public function getConfigTreeBuilder()
                         ->end()
                     ->end()
                 ->end()
+                ->scalarNode('serializer_adapter')->end()
```
Comment (csarrazi) [inline src/DependencyInjection/Configuration.php]: I would rather have used an arraynode. See how the cache node is handled.

File: src/GuzzleHttp/Cache/DoctrineAdapter.php [hunk 7]
```diff
@@ -54,8 +54,8 @@ public function save(RequestInterface $request, ResponseInterface $response)
     private function getKey(RequestInterface $request)
     {
         return md5(serialize([
-            'method'  => $request->getMethod(),
-            'uri'     => $request->getUri(),
```
Comment (csarrazi) [inline src/GuzzleHttp/Cache/DoctrineAdapter.php]: Same thing. CS will be part of another PR.

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+     *
+     * @param array $options
+     *
+     * @return array
+     */
+    private function prepareDefaults(array $options = [])
+    {
+        $serializerAdapter = $this->getConfig('serializer_adapter');
+
+        if ((isset($options['json'])) && $serializerAdapter) {
```
Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: I would also check whether the data is an array or not. In this case, we simply return from the function.

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            $modify['body'] = Psr7\stream_for($serializerAdapter->serialize($options['json'], 'json', []));
+            $options['_conditional']['Content-Type'] = 'application/json';
+            unset($options['json']);
```
Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: Add a `return` here. I don't think it makes sense to try serializing in XML after already having a json. Or throw an exception elsewhere.

You should check the client's default behavior in this case.

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+     *
+     * @return mixed
+     */
+    private function processResponse(Psr7\Response $response)
+    {
+        $serializerAdapter = $this->getConfig('serializer_adapter');
+
+        if ($response->hasHeader('Content-Type') && $serializerAdapter) {
+            $format = false;
+            if (in_array('application/json', $response->getHeader('Content-Type'))) {
```
Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: For this part, we should instead check whether the header contains: `application/*+json`, `application/json` or `text/javascript`.

Same thing for XML below: `application/*+xml`, `text/xml`, `application/xml`.

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                $format = 'json';
+            } elseif (in_array('application/xml', $response->getHeader('Content-Type'))) {
+                $format = 'xml';
+            }
+
+            if ($format) {
+
```
Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: Remove empty line.

File: src/DependencyInjection/Configuration.php (hunk 6 above)
Comment (csarrazi) [inline src/DependencyInjection/Configuration.php]: This approach would also simplify configuration, by making the serializer entirely optional.

**Comment by csarrazi** [comment]
Don't forget to add `symfony/serializer` and `jms/serializer` in the `suggests` and `require-dev` nodes in composer.json.

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+     * @param array $options
+     *
+     * @return array
+     */
+    private function prepareDefaults(array $options = [])
+    {
+        $serializerAdapter = $this->getConfig('serializer_adapter');
+
+        if ((isset($options['json'])) && $serializerAdapter) {
+            $modify['body'] = Psr7\stream_for($serializerAdapter->serialize($options['json'], 'json', []));
```
Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: This won't work for Guzzle 4 or 5. I guess I'll cherry-pick your commits to merge them into `1.3`, and then into `master`. Ok ? :)

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+     * @param array $options
+     *
+     * @return array
+     */
+    private function prepareDefaults(array $options = [])
+    {
+        $serializerAdapter = $this->getConfig('serializer_adapter');
+
+        if ((isset($options['json'])) && $serializerAdapter) {
+            $modify['body'] = Psr7\stream_for($serializerAdapter->serialize(
```
Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: This won't work for Guzzle 4 or 5. I guess I'll cherry-pick your commits to merge them into `1.3`, and then into `master`. Ok ? :)

File: src/GuzzleHttp/ClientSerializerAdapter.php (hunk 8 above)
Comment (florianbelhomme) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: Array can also be serialized by JMS & cie for childrens

Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: Ok. Let's keep it as is, then.

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+use GuzzleHttp\Client;
+use Psr\Http\Message\RequestInterface;
+
+class ClientSerializerAdapter extends Client
+{
+    /**
+     * @var string
+     */
+    private $jsonRegex = '((application|text)/([^+]*\+){0,1}json)';
```
Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: This won't match with `text/javascript`, which is another content-type which may be used for returning json.

File: src/Serializer/SymfonySerializerAdapter.php [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function serialize($data, $format, $context = [])
+    {
+        if (!is_array($context)) {
+            throw new \RuntimeException(
+                'Serialize context must be an array.'
```
Comment (csarrazi) [inline src/Serializer/SymfonySerializerAdapter.php]: Serialization context

File: src/Serializer/SymfonySerializerAdapter.php [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function deserialize($data, $type, $format, $context = [])
+    {
+        if (!is_array($context)) {
+            throw new \RuntimeException(
+                'Deserialize context must be an array.'
```
Comment (csarrazi) [inline src/Serializer/SymfonySerializerAdapter.php]: Deserialization context

File: src/GuzzleHttp/ClientSerializerAdapter.php (hunk 14 above)
Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: You should use the following expression: `(text/javascript)|(application/([^+]*\+){0,1}json)`.

File: src/Serializer/SymfonySerializerAdapter.php [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+     * @var SymfonySerializer
+     */
+    protected $serializer;
+
+    /**
+     * @param SymfonySerializer $serializer
+     */
+    public function __construct(SymfonySerializer $serializer)
+    {
+        $this->$serializer = $serializer;
```
Comment (salahm) [inline src/Serializer/SymfonySerializerAdapter.php]: `$this->serializer`

File: src/Serializer/SymfonySerializerAdapter.php [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+{
+    /**
+     * @var SymfonySerializer
+     */
+    protected $serializer;
+
+    /**
+     * @param SymfonySerializer $serializer
+     */
+    public function __construct(SymfonySerializer $serializer)
```
Comment (salahm) [inline src/Serializer/SymfonySerializerAdapter.php]: i'm really not fan of using aliases in typehinting, they are really confusing

File: src/Serializer/SymfonySerializerAdapter.php [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $this->$serializer = $serializer;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function serialize($data, $format, $context = [])
+    {
+        if (!is_array($context)) {
```
Comment (salahm) [inline src/Serializer/SymfonySerializerAdapter.php]: you can just type hint the context here ;) 
`public function serialize($data, $format,array $context = [])`

File: src/Serializer/SymfonySerializerAdapter.php [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+        return $this->serializer->serialize($data, $format, $context);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function deserialize($data, $type, $format, $context = [])
+    {
+        if (!is_array($context)) {
```
Comment (salahm) [inline src/Serializer/SymfonySerializerAdapter.php]: same here

File: src/Serializer/SymfonySerializerAdapter.php (hunk 18 above)
Comment (salahm) [inline src/Serializer/SymfonySerializerAdapter.php]: i think, you should rely on an Interface rather than an implementation.

File: src/Serializer/JMSSerializerAdapter.php [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+        return $this->serializer->serialize($data, $format, $context);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function deserialize($data, $type, $format, $context = null)
+    {
+        if (!empty($context)) {
```
Comment (salahm) [inline src/Serializer/JMSSerializerAdapter.php]:
```php
$context = $context ?: JMSDeserializationContext::create();
if (!$context instanceof JMSDeserializationContext) {
                throw new \LogicException(
                    'Deserialization context must be an instance of JMS\Serializer\DeserializationContext, %s given.',
                    get_class($context)
                );
           }
```

File: src/Resources/config/serializer.xml [hunk 22]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" ?>
+
+<container xmlns="http://symfony.com/schema/dic/services"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+
+        <service id="csa_guzzle.serializer.adapter.jms" class="Csa\Bundle\GuzzleBundle\Serializer\JMSSerializerAdapter">
```
Comment (salahm) [inline src/Resources/config/serializer.xml]: should be private

File: src/Resources/config/serializer.xml [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+            <argument type="service" id="jms_serializer" />
+        </service>
+
+        <service id="csa_guzzle.serializer.adapter.symfony"
```
Comment (salahm) [inline src/Resources/config/serializer.xml]: same here

File: src/Resources/doc/specific_serializer.md [hunk 24]
```diff
@@ -0,0 +1,62 @@
+Using specific serializer
+=========================
+
+By default Guzzle allow you to set a specific variable in the JSON configuration of a rezquest.
```
Comment (iamluc) [inline src/Resources/doc/specific_serializer.md]: allow --> allows ?
rezquest --> request

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+                    $options[$format],
+                    $format,
+                    (!empty($options['serialization']['serialize_context'])) ? $options['serialization']['serialize_context'] : null
+                );
+                if (function_exists('\GuzzleHttp\Psr7\stream_for')) {
+                    $content = \GuzzleHttp\Psr7\stream_for($content);
+                }
+                $options['body'] = $content;
+                $options['_conditional']['Content-Type'] = $contentType;
+                unset($options['xml']);
```
Comment (iamluc) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: If you don't unset `$options['json']`, are you sure you don't have a double encoding (the second time by the guzzle client) ?

File: src/DependencyInjection/Configuration.php [hunk 26]
```diff
@@ -118,6 +119,7 @@ private function createClientsNode()
                     ->arrayNode('middleware')
                         ->prototype('scalar')->end()
                     ->end()
+                    ->append($this->createSerializerNode())
```
Comment (csarrazi) [inline src/DependencyInjection/Configuration.php]: Do we actually need a specific serializer for different clients ?

Regardless, generally the configuration / mapping is still the same ?

Comment (florianbelhomme) [inline src/DependencyInjection/Configuration.php]: As you will call different API, you may want different behaviours.

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+
+use GuzzleHttp\Client;
+use Psr\Http\Message\RequestInterface;
+
+class ClientSerializerAdapter extends Client
+{
+    /**
+     * @var string
+     */
+    private $jsonRegex = '(text/javascript)|(application/([^+]*\+){0,1}json)';
```
Comment (SofHad) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: why not const ?

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    /**
+     * @param string $method
+     * @param null|string $uri
+     * @param array $options
+     *
+     * @return mixed
+     */
+    public function requestAsync($method, $uri = null, array $options = [])
```
Comment (SofHad) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: It is compatible with guzzle magic methods ?

It is compatible with magic methods

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+     */
+    private $xmlRegex = 'application/([^+]*\+){0,1}xml';
+
+    /**
+     * @param RequestInterface $request
+     * @param array $options
+     *
+     * @return mixed
+     */
+    public function sendAsync(RequestInterface $request, array $options = [])
```
Comment (SofHad) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: I am not convinced that changing the client behavior is the best solution.

File: src/GuzzleHttp/ClientSerializerAdapter.php [hunk 30]
```diff
@@ -0,0 +1,141 @@
...
+     */
+    private function processResponse($response, $options = [])
+    {
+        $serializerAdapter = $this->getConfig('serializer_adapter');
+
+        if ($response->hasHeader('Content-Type') && $serializerAdapter) {
+            $contentType = $response->getHeader('Content-Type')[0];
+
+            $format = false;
+            if (preg_match('#'.$this->jsonRegex.'#', $contentType)) {
```
Comment (SofHad) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: Or you can use ```Symfony\Component\HttpFoundation\File\MimeType\MimeTypeExtensionGuesser``` to guess the type

File: src/GuzzleHttp/ClientSerializerAdapter.php (hunk 29 above)
Comment (csarrazi) [inline src/GuzzleHttp/ClientSerializerAdapter.php]: True. This is why this PR is on the waiting list while guzzle/command and guzzle/guzzle-services are migrating to Guzzle 6.0

**Comment by csarrazi** [comment]
Don't worry, I haven't forgotten you! I'm just waiting for https://github.com/guzzle/command/pull/30 in order to simplify the implementation.

**Comment by csarrazi** [comment]
@florianbelhomme I'm closing your PR for now. The next step will be done in PR #138 :)
